plugins/out_azure_kusto/azure_kusto.c

```diff
@@ -18,61 +18,266 @@
+ *     stored in the context.
+ *   - -1 on failure, indicating an error occurred during the token retrieval
+ *     process.
+ *
+ * The function performs the following steps:
+ *   1. Logs the start of the token retrieval process.
+ *   2. Checks if IMDS-based token retrieval is enabled:
+ *      - If enabled, it attempts to retrieve a token from the Azure IMDS.
+ *      - Parses the JSON response to extract the access token.
+ *      - Logs and handles errors if token extraction fails.
+ *   3. If IMDS is not used, it falls back to the client secret flow:
+ *      - Clears any existing OAuth2 payload data.
+ *      - Appends necessary parameters to the OAuth2 payload, including
+ *        grant type, scope, client ID, and client secret.
+ *      - Requests an OAuth2 token using the constructed payload.
+ *      - Logs and handles errors if token retrieval fails.
+ *   4. Logs the successful completion of the token retrieval process.
+ *
+ * Note:
+ *   - The function assumes that the context (ctx) is properly initialized
+ *     and contains valid configuration data.
+ *   - The extracted access token is stored in the context's OAuth2 structure
+ *     for subsequent use in authenticated requests.
+ */
+
+static int azure_kusto_get_oauth2_token(struct flb_azure_kusto *ctx)
+{
+    char *token = NULL;
+    int ret;
+    flb_sds_t response = NULL;
+    char *access_token = NULL;
+
+    flb_plg_debug(ctx->ins, "Starting token retrieval process");
+
+    /* Check if IMDS-based token retrieval is needed */
+    if (ctx->use_imds == FLB_TRUE) {
+        flb_plg_debug(ctx->ins, "Using IMDS for token retrieval");
+
+        response = flb_azure_imds_get_token(ctx, ctx->client_id);
+        if (!response) {
+            flb_plg_error(ctx->ins, "Failed to retrieve token from Azure IMDS");
+            return -1;
+        }
+
+        // Parse the JSON response to extract the access token
+        size_t access_token_len = 0;
+        char *p = strstr(response, "\"access_token\":\"");
+        if (p) {
+            p += strlen("\"access_token\":\"");
+            char *end = strstr(p, "\"");
+            if (end) {
+                access_token_len = end - p;
+                access_token = flb_malloc(access_token_len + 1);
+                if (access_token) {
+                    strncpy(access_token, p, access_token_len);
+                    access_token[access_token_len] = '\0';
+                }
+            }
+        }
+
+        flb_plg_debug(ctx->ins, "Access token extracted is %s", access_token);
+
+        if (!access_token) {
+            flb_plg_error(ctx->ins, "Error extracting access token from IMDS response");
+            flb_sds_destroy(response);
+            return -1;
+        }
+
+        ctx->o->access_token = flb_sds_create(access_token);
+        ctx->o->token_type = flb_sds_create("Bearer");
+        flb_free(access_token);
+
+        flb_sds_destroy(response);
+
+    } else {
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions